Draft JOSS article about SciPy elementwise iterative methods (#117)
mdhaber/scipy: paper.md

---
title: 'Vectorized, Python Array API Standard Compatible Functions for Quadrature, Series Summmation, Differentiation, Optimization, and Root Finding in SciPy'
tags:
  - Python
  - SciPy
  - integration
  - differentiation
  - optimization
  - root finding
authors:
  - name: Matt Haberland^[Corresponding author]
    affiliation: 1
    orcid: 0000-0003-4806-3601
  - name: Albert Steppi
    affiliation: 2
    orcid: 0000-0001-5871-6245
  - name: Pamphile Roy
    affiliation: 3
    orcid: 0000-0001-9816-1416
  - name: Jake Bowhay
    affiliation: 4
    orcid: 0009-0002-9559-4114
affiliations:
 - name: California Polytechnic State University, San Luis Obispo, USA
   index: 1
 - name: Quansight Labs, Austin, USA
   index: 2
 - name: Consulting Manao GMBH, Vienna, Austria
   index: 3
 - name: University of Bristol, Bristol, United Kingdom
   index: 4
date: 1 March 2025
bibliography: paper.bib

---

# Summary

> Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. When the related problems are all independent, embarrassingly parallel calculation is possible However, SciPy [@scipy] – the de facto standard Python library for such algorithms – did not offer features for vectorized solution of such problems with NumPy [@numpy] arrays, let alone array library agnostic implementations.

Reviewer: ```suggestion
Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. When the related problems are all independent, embarrassingly parallel calculation is possible. However, SciPy [@scipy] – the de facto standard Python library for such algorithms – did not offer features for vectorized solutions of such problems with NumPy [@numpy] arrays, let alone array library agnostic implementations.
```

Author: Period was needed, but I think "solution" is arguably correct - I think I'm using it as a verb, not a plural noun?

Reviewer: Ah ok I read it as there are plural problems to be solved so I was expecting plural solutions but I'm not the best grammar person so maybe best to get another opinion!

Author: ```suggestion
Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. When the related problems are all independent, embarrassingly parallel calculation is possible. However, SciPy [@scipy] – the de facto standard Python library for such algorithms – did not offer features for vectorized solution of such problems with NumPy [@numpy] arrays, let alone array library agnostic implementations.
```

> Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. When the related problems are all independent, embarrassingly parallel calculation is possible However, SciPy [@scipy] – the de facto standard Python library for such algorithms – did not offer features for vectorized solution of such problems with NumPy [@numpy] arrays, let alone array library agnostic implementations.

Reviewer 2: It may be a little confusing that we're blurring the lines between vectorization in the sense of array computing, i.e. having a high level API in terms of array operations, with underlying computations performed in efficient native code, vs the way it's used more traditionally in computing, stuff like SIMD, parallel operations on entire arrays. How about something that's more explicit about the difference, like this?

```suggestion
Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. In array computing, operations on batches of values can be vectorized. With NumPy [@numpy], operations on arrays expressed in Python can be evaluated as (sequential) loops in efficient native code, or in some cases in parallel with SIMD [@nep38]. Other Python array libraries such as CuPy [@cupy] and PyTorch [@pytorch] are able to exploit GPUs to parallelize vectorized computations. However, SciPy [@scipy] – the de facto standard Python library for solving the above problems – did not offer features to vectorize the solution process with NumPy [@numpy] arrays, let alone array library agnostic implementations.
```

Author: Looks great; I only removed JAX because these routines currently use assignment and don't work with JAX. That may change soon, but I'm not sure if we should mention it before it's implemented. Thoughts about that, @steppi?

Reviewer 2: I was thinking it might still be good to mention JAX here  just to imply that it's something we're considering and to not give the impression that we're leaving out or are unaware of a major player. Then mention later on that it *currently* supports CuPy and PyTorch. There are good reasons why JAX isn't supported yet, and if anything, it would be good to raise more awareness of this.

Author: OK, done. LMK how it looks.

> Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. In array computing, operations on batches of values can be vectorized. With NumPy [@numpy], operations on arrays expressed in Python can be evaluated as (sequential) loops in efficient native code, or in some cases in parallel with SIMD [@nep38]. Other Python array libraries such as CuPy [@cupy] and PyTorch [@pytorch] are able to exploit GPUs to parallelize vectorized computations. However, SciPy [@scipy] – the de facto standard Python library for solving the above problems – did not offer features to vectorize the solution process with NumPy [@numpy] arrays, let alone array library agnostic implementations.

Author: ```suggestion
Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. In array computing, operations on batches of values can be vectorized. With NumPy [@numpy], operations on arrays expressed in Python can be evaluated as (sequential) loops in efficient native code, or in some cases in parallel with SIMD [@nep38]. Other Python array libraries such as CuPy [@cupy], PyTorch [@pytorch], and JAX [@jax] are able to exploit GPUs to parallelize vectorized computations. However, SciPy [@scipy] – the de facto standard Python library for solving the above problems – did not offer features to vectorize the solution process with NumPy [@numpy] arrays, let alone array library agnostic implementations.
```

> Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. In array computing, operations on batches of values can be vectorized. With NumPy [@numpy], operations on arrays expressed in Python can be evaluated as (sequential) loops in efficient native code, or in some cases in parallel with SIMD [@nep38]. Other Python array libraries such as CuPy [@cupy], PyTorch [@pytorch], and JAX [@jax] are able to exploit GPUs to parallelize vectorized computations. However, SciPy [@scipy] – the de facto standard Python library for solving the above problems – did not offer features to vectorize the solution process with NumPy [@numpy] arrays, let alone array library agnostic implementations.

Author: > did not offer features to vectorize the solution process with NumPy

Technically, `newton` is vectorized. Thoughts on how to adjust this sentence to be more compatible with this fact without getting awkward?

> let alone array library agnostic implementations.

I think I'll change this to "let alone alternative array libraries."

Reviewer 2: ```suggestion
Numerical integration, series summation, differentiation, optimization, and root finding are fundamental problems with applications in essentially all domains of science and engineering. Frequently, such problems do not arise individually, but rather in batches; e.g., differentiation of a single curve at many points or minimization of a function for many values of a parameter. In array computing, operations on batches of values can be vectorized. With NumPy [@numpy], operations on arrays expressed in Python can be evaluated as (sequential) loops in efficient native code, or in some cases in parallel with SIMD [@nep38]. Other Python array libraries such as CuPy [@cupy], PyTorch [@pytorch], and JAX [@jax] are able to exploit GPUs to parallelize vectorized computations. However, SciPy [@scipy] – the de facto standard Python library for solving the above problems – offered few functions capable of vectorizing the solution of such problems with NumPy [@numpy], let alone with alternative array libraries.
```
Maybe something like this?

Author: That's similar to what I was thinking.
How about just "few functions"? "scarcely any..." is a bet less common and might draw attention to itself, whereas this is a detail that just gets in the way and should be ignored.

Reviewer 2: "few" sounds good to me

This paper discusses several new features that fill this gap, included in SciPy 1.15.0:

- `scipy.differentiate`, a sub-package for numerical differentiation of scalar or vector-valued functions of one or more variables,
- `scipy.integrate.tanhsinh` for quadrature of scalar or vector-valued integrands of one variable,
- `scipy.integrate.nsum` for summation of real-valued finite or infinite series,

> - `scipy.optimize.elementwise` for finding roots and minimization of a single-input, single output function.

Reviewer: ```suggestion
- `scipy.optimize.elementwise` for finding roots and minimization of a single-input, single-output function.
```
For consistency with single-input

> - `scipy.optimize.elementwise` for finding roots and minimization of a single-input, single-output function.

Author: ```suggestion
- `scipy.optimize.elementwise` for root finding and minimization of a single-input, single-output function.
```

> Although the details of the algorithms are inherently distinct, these features rely on a common framework for elementwise iterative methods and share similar interfaces, and the same implementation works with several Python Array API Standard [@arrayapi] compatible arrays. These features will dramatically improve performance of end-user applications, and together, they form the backbone of SciPy's new random variable infrastructure.

Author: ```suggestion
Although the details of the algorithms are inherently distinct, these features rely on a common framework for elementwise iterative methods and share similar interfaces, and the same implementation works with several Python Array API Standard [@arrayapi] compatible arrays, including CuPy and PyTorch. These features will dramatically improve performance of end-user applications, and together, they form the backbone of SciPy's new random variable infrastructure.
```

> Although the details of the algorithms are inherently distinct, these features rely on a common framework for elementwise iterative methods and share similar interfaces. For example, `scipy.differentiate.derivative` accepts the scalar function to be differentiated as a Python callable, the points at which to evaluate the derivative as an N-dimensional array, and any other parameters of the callable as N-dimensional arrays. `derivative` passes abscissae and parameters to the callable as N-dimensional arrays, adaptively approximating the derivative(s) to achieve specified tolerances. Furthermore, arrays need not be NumPy arrays: the same implementation works with several Python Array API Standard [@arrayapi] compatible arrays, such as PyTorch [@pytorch] tensors and CuPy [@cupy] arrays. These features will dramatically improve performance of end-user applications, and together, they form the backbone of SciPy's new random variable infrastructure.

Author: ```suggestion
Although the details of the algorithms are inherently distinct, these features rely on a common framework for elementwise iterative methods and share similar interfaces. For example, `scipy.differentiate.derivative` accepts the scalar function to be differentiated as a Python callable, the points at which to evaluate the derivative as an N-dimensional array, and any other parameters of the callable as N-dimensional arrays. `derivative` passes abscissae and parameters to the callable as N-dimensional arrays, adaptively approximating the derivative(s) to achieve specified tolerances. Furthermore, arrays need not be NumPy arrays: the same implementation works with several Python Array API Standard [@arrayapi] compatible arrays. These features will dramatically improve performance of end-user applications, and together, they form the backbone of SciPy's new random variable infrastructure.
```

> Although the details of the algorithms are inherently distinct, these features rely on a common framework for elementwise iterative methods and share similar interfaces. For example, `scipy.differentiate.derivative` accepts the scalar function to be differentiated as a Python callable, the points at which to evaluate the derivative as an N-dimensional array, and any other parameters of the callable as N-dimensional arrays. `derivative` passes abscissae and parameters to the callable as N-dimensional arrays, adaptively approximating the derivative(s) to achieve specified tolerances. Furthermore, arrays need not be NumPy arrays: the same implementation works with several Python Array API Standard [@arrayapi] compatible arrays. These features will dramatically improve performance of end-user applications, and together, they form the backbone of SciPy's new random variable infrastructure.

Author: @tupui I think you were suggesting that this example was hard to read and suggested that I show code instead of all the words? I don't want to show code because I want to stay very far from the guideline that "API (Application Programming Interface) functionality should not be in the paper and instead should be outlined in the software documentation". I thought about just removing it:

```suggestion
Although the details of the algorithms are inherently distinct, these features rely on a common framework for elementwise iterative methods and share similar interfaces, and the same implementation works with several Python Array API Standard [@arrayapi] compatible arrays. These features will dramatically improve performance of end-user applications, and together, they form the backbone of SciPy's new random variable infrastructure.
```

Thoughts?

Reviewer 2: I support taking out this part. It seems more in line with the [JOSS guidelines](https://joss.readthedocs.io/en/latest/paper.html#what-should-my-paper-contain) to leave these kinds of details out.

Reviewer 3: Sounds good 👍

# Statement of need

> Before the release of SciPy 1.15.0, the need for these capabilities was partially met in the scientific Python ecosystem. As popular examples, Numdifftools [@numdifftools] and PyNumDiff [@PyNumDiff] provide tools for numerical differentiation, `quadpy` [@quadpy] and `scipy.integrate` offer several functions for numerical integration, `mpmath`'s [@mpmath] `nsum` function is for series summation, and `scipy.optimize` offers functions for scalar minimization and root finding. However, to the authors' knowledge, the new implementations are unique in that they offer the following advantages.

Reviewer: Fyi quadpy is now closed source https://github.com/sigma-py/quadpy/issues/481

Author: Yeah. I saw but wasn't sure what to write about that. It still seems reasonable to cite.

> - Backend-independence: all of these features are "Array API compatible" in the sense that they rely almost entirely on the Python standard libary and calls to Python Array API Standard functions/methods. Consequently, the functions accept objects other than NumPy arrays, such as PyTorch tensors and CuPy arrays, and the calculations are performed by the underlying array library.

Reviewer: ```suggestion
- Backend-independence: all of these features are "Array API compatible" in the sense that they rely almost entirely on the Python standard library and calls to Python Array API Standard functions/methods. Consequently, the functions accept objects other than NumPy arrays, such as PyTorch tensors and CuPy arrays, and the calculations are performed by the underlying array library.
```

Author: I should probably add a spell checker plugin to my text editor : )

- Speed: the features take full advantage of vectorized user callables, avoiding slow Python loops and the excessive overhead of repeatedly calling compiled code. 
- Prevalence: SciPy is one of the most popular scientific Python packages. If a scientific Python user needs these features, chances are that they already have SciPy installed, eliminating the need to find and learn a new package.

> - Ease-of-use: the function API reference is thorough, and the interfaces share common features and operate smoothly with other SciPy functions.

Reviewer 2: ```suggestion
- Ease-of-use: the API reference for these new functions is thorough, and the interfaces share common features and operate smoothly with other SciPy functions.
```

Perhaps this is on me, but it it took me three seconds to realize I should mentally parse this as "the function (API reference)" instead of "the (function API) reference" and in the context of "Array API" I was wondering what the "function API" is.

> - Dependability: as with all *new* SciPy code, these features were designed and implemented following good software development practices. They have been carefully peer-reviewed, and extensive unit tests protect against backward incompatible changes and regressions.

Author: ```suggestion
- Dependability: as with all new SciPy code, these features were designed and implemented following good software development practices. They have been carefully peer-reviewed, and extensive unit tests protect against backward incompatible changes and regressions.
```

> - Dependability: as with all SciPy code, these features were designed and implemented following good software development practices. They have been carefully peer-reviewed, and extensive unit tests protect against backward incompatible changes and regressions.

Reviewer 2: ```suggestion
- Dependability: as with all *new* SciPy code, these features were designed and implemented following good software development practices. They have been carefully peer-reviewed, and extensive unit tests protect against backward incompatible changes and regressions.
```

This could be triggering to someone who's been mired in the depths of some of the legacy stuff that's floating around.

Author: I actually thought about adding exactly this while investigating mstats.
